src/Language/PureScript/Linter.hs

```diff
@@ -174,113 +174,116 @@ lintUnused (Module modSS _ mn modDecls exports) =
+            (vars, errs) = removeAndWarn bindNewNames $ mconcat $ map go allExprs
             errs' = addHint (ErrorInValueDeclaration $ valdeclIdent vd) errs
         in
           (vars, errs')
 
     goDecl (TypeInstanceDeclaration _ _ _ _ _ _ _ (ExplicitInstance decls)) = mconcat $ map goDecl decls
     goDecl _ = mempty
 
-    go :: SourceSpan -> Expr -> (S.Set Ident, MultipleErrors)
-    go _ (Var _ (Qualified Nothing v)) = (S.singleton v, mempty)
-    go _ (Var _ _) = (S.empty, mempty)
-
-    go ss (Let _ ds e) =
-      let letNames = S.fromList $ concatMap declIdents ds
-      in removeAndWarn ss letNames $ mconcat (go ss e : map underDecl ds)
-    go ss (Abs binder v1) =
-      let newNames = S.fromList (binderNames binder)
+    go :: Expr -> (S.Set Ident, MultipleErrors)
+    go (Var _ (Qualified Nothing v)) = (S.singleton v, mempty)
+    go (Var _ _) = (S.empty, mempty)
+
+    go (Let _ ds e) =
+      let (letNames, letNamesRec) = foldMap declIdents ds
+      in removeAndWarn letNamesRec $
+            removeAndWarn letNames (go e)
+            <> mconcat (map underDecl ds)
+    go (Abs binder v1) =
+      let newNames = S.fromList (binderNamesWithSpans binder)
       in
-      removeAndWarn ss newNames $ go ss v1
+      removeAndWarn newNames $ go v1
 
-    go ss (UnaryMinus _ v1) = go ss v1
-    go ss (BinaryNoParens v0 v1 v2) = go ss v0 <> go ss v1 <> go ss v2
-    go ss (Parens v1) = go ss v1
-    go ss (TypeClassDictionaryConstructorApp _ v1) = go ss v1
-    go ss (Accessor _ v1) = go ss v1
+    go (UnaryMinus _ v1) = go v1
+    go (BinaryNoParens v0 v1 v2) = go v0 <> go v1 <> go v2
+    go (Parens v1) = go v1
+    go (TypeClassDictionaryConstructorApp _ v1) = go v1
+    go (Accessor _ v1) = go v1
 
-    go ss (ObjectUpdate obj vs) = mconcat (go ss obj : map (go ss . snd) vs)
-    go ss (ObjectUpdateNested obj vs) = go ss obj <> goTree vs
+    go (ObjectUpdate obj vs) = mconcat (go obj : map (go . snd) vs)
+    go (ObjectUpdateNested obj vs) = go obj <> goTree vs
       where
         goTree (PathTree tree) = mconcat $ map (goNode . snd) (runAssocList tree)
-        goNode (Leaf val) = go ss val
+        goNode (Leaf val) = go val
         goNode (Branch val) = goTree val
 
-    go ss (App v1 v2) = go ss v1 <> go ss v2
-    go ss (Unused v) = go ss v
-    go ss (IfThenElse v1 v2 v3) = go ss v1 <> go ss v2 <> go ss v3
-    go ss (Case vs alts) =
+    go (App v1 v2) = go v1 <> go v2
+    go (Unused v) = go v
+    go (IfThenElse v1 v2 v3) = go v1 <> go v2 <> go v3
+    go (Case vs alts) =
       let f (CaseAlternative binders gexprs) =
-            let bindNewNames = S.fromList (concatMap binderNames binders)
+            let bindNewNames = S.fromList (concatMap binderNamesWithSpans binders)
                 allExprs = concatMap unguard gexprs
             in
-                removeAndWarn ss bindNewNames $ mconcat $ map (go ss) allExprs
+                removeAndWarn bindNewNames $ mconcat $ map (go) allExprs
```

Comment:
Unneeded parens in `(go)`. (I found four of these in total; they're easy to search for.)

Reply:
Oh dear, a little bit naive on the search and replace :)